content/blog/2019-12-02-equivalence-modulo-inputs/index.md

```diff
@@ -0,0 +1,400 @@
+majority (all but four) of bugs identified by EMI were found with EMI variants
+of random programs generated by Csmith.
+
+[Differential testing]: https://www.cs.swarthmore.edu/~bylvisa1/cs97/f13/Papers/DifferentialTestingForSoftware.pdf
+
+
+## Some definitions
+
+Let us now formally define EMI and show how we can use it as a condition
+to determine whether a compiler is buggy.
+
+> **Equivalence modulo inputs (EMI)**.
+  Given an input set *I* and a program *P*, a program *Q* is an EMI-variant of
+  *P* (read: *P* and *Q* are EMI) relative to *I* if *P* and *Q* have
+  the same denotations ("mean the same thing") over all inputs in *I*.
+  Formally, for all *i* in *I*, `⟦P⟧(i) = ⟦Q⟧(i)`.
+
+
+It is immediately clear that EMI is a relaxation of semantic equivalence,
+wherein *P* and *Q* have the same denotations for all possible inputs.
+
+For example, the following two programs are semantically equivalent
+(and thus EMI for any input set):
+
+> `\x. x + x` and `\x. 2 * x`
+
+The following two programs are *not* semantically equivalent yet EMI over
+input set `{0}`:
+
+> `\x. x` and `\x. 0`
+
+Now that we have a formal definition of EMI, how can we use it as a condition
+to check whether a compiler is buggy or not?
+
+> **EMI-validity**. Given an input set *I*, a compiler *C* is *EMI-valid*
+  relative to *I* if for any program *P* and EMI(I)-variant *Q*,
+  it is the case that `C(P)(i) = C(Q)(i)` for all *i* in input set I. 
+
+**If a compiler is not EMI-valid, then we consider it buggy.** 
+But the inverse is not true: if a compiler *is* EMI-valid, it can
+still be buggy!
+Consider the degenerate compiler that maps all source programs to the same
+target program.
+The compiler is EMI-valid for any input set *I*, but it is obviously buggy.
+Thus EMI-validity is a conservative overapproximation for compiler correctness.
+
+Why is this useful? Couldn't we just define validity for a compiler over
+*semantic equivalence*, not just its relaxed counterpart in EMI?
+(You can imagine defining EMI-validity where the input set is all possible
+inputs.)
+EMI solves two hard practical problems in differential testing of compilers:
+
+1. How do we generate "equivalent" variants of input programs?
+
+2. How do we check that the compiler's output programs are "equivalent"?
+
+Using the more stringent condition of semantic equivalence makes solving
+these practical problems hard&mdash;indeed, in the general case (2) is undecidable
+by a [famous result in computability theory][rice].
+But the more relaxed condition of EMI makes these tractable.
+As we'll seen with the implementation of Orion below, using EMI allows us
+an efficient procedure for generating EMI-variants from seed programs,
+thus solving (1).
+We determine that output programs are "equivalent" if they are EMI,
+which, since we're only checking equivalence over a particular set of
+inputs, gives an efficient procedure for (2).
+
+[rice]: https://en.wikipedia.org/wiki/Rice%27s_theorem
```

Comment:
Really nice work on this overview section! You've succeeded in giving a specific, clear, and complete explanation of the somewhat tricky core idea here. :sparkles: